File: cracking_the_coding_interview/17.21-volume-of-histogram-lcci.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        // 每一格子的水量, 取决其左右最高值
        const int n = height.size();
        vector<int> left(n);
        vector<int> right(n);
        int v = 0;
        for (int i = 0; i < n; ++i) {
            v = max(v, height[i]);
            left[i] = v;
        }
        v = 0;
        for (int i = n - 1; i >= 0; --i) {
            v = max(v, height[i]);
            right[i] = v;
        }
        int s = 0;
        for (int i = 0; i < n; ++i) {
            s += min(left[i], right[i]) - height[i];
        }
        return s;
    }
};

class Solution {
public:
    int trap(vector<int>& height) {
        if (height.size() <= 2) return 0;
        int left = 0, right = height.size() - 1;
        int left_most = height.front(), right_most = height.back();
        int ans = 0;
        while (left < right) {
            if (left_most < right_most) {
                ans += left_most - height[left];
                left_most = max(left_most, height[++left]);
            } else {
                ans += right_most - height[right];
                right_most = max(right_most, height[--right]);
            }
        }
        return ans;
    }
};

File: cracking_the_coding_interview/17.21-volume-of-histogram-lcci.go
package main

import "fmt"


func trap3(height []int) int {
    /*
        water[i] = min(max(height[:i]), max(height[i+1:])) - height[i] 
     */
    if len(height) == 0 {
        return 0
    }
    ans := 0
    st := make([]int, 0)
    for i, h := range height {
        for len(st) > 0 && h >= height[st[len(st) - 1]] {
            ti := st[len(st) - 1]
            st = st[:len(st) - 1]
            if len(st) == 0 {
                break
            }
            left := st[len(st) - 1]
            dh := min(height[left], height[i]) - height[ti]
            ans += dh * (i - left - 1)
        }
        st = append(st, i)
    }
    return ans
}

// 双指针
func trap(height []int) int {
    if len(height) <= 2 {
        return 0
    }
    left, right := 0, len(height) - 1
    left_max, right_max := height[left], height[right]
    ans := 0
    for left < right {
        if left_max <= right_max {
            ans += left_max - height[left]
            left += 1
            left_max = max(left_max, height[left])
        } else {
            ans += right_max - height[right]
            right -= 1
            right_max = max(right_max, height[right])
        }
    }
    return ans
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func trap2(height []int) int {
    /*
        water[i] = min(max(height[:i]), max(height[i+1:])) - height[i] 
     */
    if len(height) <= 2 {
        return 0
    }
    ans := 0
    left := make([]int, len(height))
    right := make([]int, len(height))

    left[0] = height[0]
    for i := 1; i < len(height); i++ {
        left[i] = max(left[i - 1], height[i])
    }
    right[len(height) - 1] = height[len(height) - 1]
    for i := len(height) - 2; i >= 0; i-- {
        right[i] = max(right[i + 1], height[i])
    }
    for i := 1; i < len(height) - 1; i++ {
        ans += min(left[i], right[i]) - height[i]
    }

    return ans
}


func main() {
    height := []int{0,1,0,2,1,0,1,3,2,1,2,1}
    fmt.Println(trap(height))
}
